### PR TITLE
HIVE-26481:Cleaner fails with FileNotFoundException

### PR DESCRIPTION
**Pull request by rkirtir**
### What changes were proposed in this pull request?

Handled FileNotFoudExcrption in cleaner

### Why are the changes needed?

When cleaner tries to remove files from HDFS location runs into java.io.FileNotFoundException:.In order to fix this we need to handle a exception thrown by DistributedFileSystem.listLocatedStatus method ->FileNotFoundException

### Does this PR introduce _any_ user-facing change?

No
### How was this patch tested?

With junit tests

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java [hunk 1]
```diff
@@ -1499,29 +1499,34 @@ private static ValidTxnList getValidTxnList(Configuration conf) {
...
-        Path fPath = fStatus.getPath();
-        if (acidHiddenFileFilter.accept(fPath)) {
-          if (baseFileFilter.accept(fPath) ||
-                  deltaFileFilter.accept(fPath) ||
-                  deleteEventDeltaDirFilter.accept(fPath)) {
-            addToSnapshoot(dirToSnapshots, fPath);
-          } else {
-            if (fStatus.isDirectory()) {
-              stack.push(fs.listStatusIterator(fPath));
+    try {
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1502]: can we wrap with try block as small as possible?

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java [hunk 2]
```diff
@@ -1499,29 +1499,34 @@ private static ValidTxnList getValidTxnList(Configuration conf) {
...
+                // Found an original file
+                HdfsDirSnapshot hdfsDirSnapshot = addToSnapshoot(dirToSnapshots, fPath.getParent());
+                hdfsDirSnapshot.addFile(fStatus);
+              }
             }
           }
         }
       }
+    } catch (FileNotFoundException fne) {
+      //ignore
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1528]: please add some comment that the current FS API doesn't provide the ability to supply a PathFilter to ignore the staging dirs, and that is why we need to catch this exception

Comment (rkirtir) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1528]: incorporated this comment

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java [hunk 3]
```diff
@@ -1503,24 +1503,31 @@ public static Map<Path, HdfsDirSnapshot> getHdfsDirSnapshotsForCleaner(final Fil
...
+                HdfsDirSnapshot hdfsDirSnapshot = addToSnapshoot(dirToSnapshots, fPath.getParent());
+                hdfsDirSnapshot.addFile(fStatus);
+              }
             }
           }
         }
+      }catch(FileNotFoundException fne){
+        //Ignore....
+        //As current FS API doesn't provide the ability to supply a PathFilter to ignore the staging dirs,
+        // need to catch this exception
```
Comment (ayushtkn) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1529]: There is something like RemoteIterators.filteringRemoteIterator(fs.listStatusIterator(path), <filter Condition>) provided by Hadoop for such use cases.

Comment (rkirtir) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java]: Thanks Ayush for this. I have made necessary changes with RemoteIterators.filteringRemoteIterator

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java [hunk 4]
```diff
@@ -1499,10 +1500,10 @@ private static ValidTxnList getValidTxnList(Configuration conf) {
   public static Map<Path, HdfsDirSnapshot> getHdfsDirSnapshotsForCleaner(final FileSystem fs, final Path path)
           throws IOException {
     Map<Path, HdfsDirSnapshot> dirToSnapshots = new HashMap<>();
-    Deque<RemoteIterator<FileStatus>> stack = new ArrayDeque<>();
-    stack.push(fs.listStatusIterator(path));
+    Deque<RemoteIterator<LocatedFileStatus>> stack = new ArrayDeque<>();
+    stack.push(FileUtils.listAndFilter (fs, path, true, acidHiddenFileFilter));
```
Comment (ayushtkn) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1504]: Earlier it was listStatusIterator, you changed it to listFiles with recursive, it won't be having directories now, empty directories and stuff like that will get completely skipped, I am not sure if you can do that. Lets not change the behaviour in this scope.

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/utils/FileUtils.java [hunk 5]
```diff
@@ -654,4 +655,11 @@ void findNext() throws IOException {
       nextFile = null;
     }
   }
+
+  public static RemoteIterator<LocatedFileStatus> listAndFilter(FileSystem fileSystem, Path path, boolean recursive,
+                                                                PathFilter filter) throws IOException {
+    return  RemoteIterators.filteringRemoteIterator(
+            fileSystem.listFiles(path, recursive),
+            status -> filter.accept(status.getPath()));
+  }
```
Comment (ayushtkn) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/utils/FileUtils.java:664]: Code isn't formatted. Use `eclipse-styles.xml` present in dev-support in hive repo to format the code

**Review by ayushtkn — COMMENTED** [review]
If possible we should extend a test for this use case as well, or atleast for the util added to check if it behaves in the way we want to

File: common/src/test/org/apache/hadoop/hive/common/TestFileUtils.java [hunk 6]
```diff
@@ -35,14 +35,12 @@
 import java.util.HashSet;
 import java.util.Set;
 
-import org.apache.hadoop.fs.ContentSummary;
-import org.apache.hadoop.fs.FileSystem;
-import org.apache.hadoop.fs.LocalFileSystem;
-import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.*;
```
Comment (ayushtkn) [inline common/src/test/org/apache/hadoop/hive/common/TestFileUtils.java:38]: don't club the imports.

File: common/src/test/org/apache/hadoop/hive/common/TestFileUtils.java [hunk 7]
```diff
@@ -281,4 +279,32 @@ public void testMakeRelative() {
...
+    try {
+      int count = 0;
+      while (lfs.hasNext()) {
+        lfs.next();
+        count++;
+      }
+      return count;
+    } catch (IOException e) {
+      throw new Exception("Exception while list files on " + e);
+    }
```
Comment (ayushtkn) [inline common/src/test/org/apache/hadoop/hive/common/TestFileUtils.java:307]: there is no need to catch & throw let the exception surface

File: common/src/test/org/apache/hadoop/hive/common/TestFileUtils.java [hunk 8]
```diff
@@ -281,4 +279,32 @@ public void testMakeRelative() {
...
+            new MockFileSystem.MockFile("mock:/tmp/_dummy", 500, new byte[0]),
+            new MockFileSystem.MockFile("mock:/tmp/dummy", 500, new byte[0]));
+    Path path = new MockFileSystem.MockPath(fs, "/tmp");
+    RemoteIterator<FileStatus> it = FileUtils.listStatusIterator(fs, path, FileUtils.HIDDEN_FILES_PATH_FILTER);
+    Assert.assertEquals(1, getFileCount(it));
+    RemoteIterator<LocatedFileStatus> itr = FileUtils.listFilesIterator(fs, path, true, FileUtils.HIDDEN_FILES_PATH_FILTER);
+    Assert.assertEquals(1, getFileCount(itr));
+  }
+
+  private int getFileCount(RemoteIterator<?> lfs) throws Exception {
```
Comment (ayushtkn) [inline common/src/test/org/apache/hadoop/hive/common/TestFileUtils.java:297]: ? can be replaced by `? extends FileStatus`, LocatedFileStatus also extends from `FileStatus`

File: common/src/test/org/apache/hadoop/hive/common/TestFileUtils.java [hunk 9]
```diff
@@ -281,4 +279,32 @@ public void testMakeRelative() {
...
+
+    MockFileSystem fs = new MockFileSystem(new HiveConf(),
+            new MockFileSystem.MockFile("mock:/tmp/.staging", 500, new byte[0]),
+            new MockFileSystem.MockFile("mock:/tmp/_dummy", 500, new byte[0]),
+            new MockFileSystem.MockFile("mock:/tmp/dummy", 500, new byte[0]));
+    Path path = new MockFileSystem.MockPath(fs, "/tmp");
+    RemoteIterator<FileStatus> it = FileUtils.listStatusIterator(fs, path, FileUtils.HIDDEN_FILES_PATH_FILTER);
+    Assert.assertEquals(1, getFileCount(it));
+    RemoteIterator<LocatedFileStatus> itr = FileUtils.listFilesIterator(fs, path, true, FileUtils.HIDDEN_FILES_PATH_FILTER);
+    Assert.assertEquals(1, getFileCount(itr));
```
Comment (ayushtkn) [inline common/src/test/org/apache/hadoop/hive/common/TestFileUtils.java:294]: there is already a static import for assertEquals in the class, no need to prefix Assert.

File: common/src/test/org/apache/hadoop/hive/common/TestFileUtils.java [hunk 10]
```diff
@@ -281,4 +279,32 @@ public void testMakeRelative() {
...
+  }
+
+  private int getFileCount(RemoteIterator<?> lfs) throws Exception {
+    try {
+      int count = 0;
+      while (lfs.hasNext()) {
+        lfs.next();
+        count++;
```
Comment (ayushtkn) [inline common/src/test/org/apache/hadoop/hive/common/TestFileUtils.java:302]: Can assert the expected file paths as well. Can change the util as:
```
  private int assertExpectedFilePaths(RemoteIterator<? extends FileStatus> lfs, List<String> expectedPaths)
      throws Exception {
    int count = 0;
    while (lfs.hasNext()) {
      assertTrue(expectedPaths.contains(lfs.next().getPath().toString()));
      count++;
    }
    return count;
  }
```

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java [hunk 11]
```diff
@@ -1500,7 +1500,7 @@ public static Map<Path, HdfsDirSnapshot> getHdfsDirSnapshotsForCleaner(final Fil
           throws IOException {
     Map<Path, HdfsDirSnapshot> dirToSnapshots = new HashMap<>();
     Deque<RemoteIterator<FileStatus>> stack = new ArrayDeque<>();
-    stack.push(fs.listStatusIterator(path));
+    stack.push(FileUtils.listStatusIterator(fs,path,acidHiddenFileFilter));
```
Comment (ayushtkn) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1503]: nit:
missing space after comma.

**Review by ayushtkn — CHANGES_REQUESTED** [review]
Dropped some comments, Prod code LGTM

**Review by ayushtkn — APPROVED** [review]
LGTM

**Review by deniskuzZ — APPROVED** [review]

**Pull request merged**